kinesis-video-c-producer/samples/KvsAacAudioVideoStreamingSample.c

```diff
@@ -0,0 +1,285 @@
+#define SAMPLE_AUDIO_FRAME_DURATION         (20 * HUNDREDS_OF_NANOS_IN_A_MILLISECOND)
+#define SAMPLE_VIDEO_FRAME_DURATION         (HUNDREDS_OF_NANOS_IN_A_SECOND / DEFAULT_FPS_VALUE)
+
+#define NUMBER_OF_H264_FRAME_FILES          403
+#define NUMBER_OF_AAC_FRAME_FILES           582
+
+typedef struct {
+    volatile ATOMIC_BOOL firstVideoFramePut;
+    UINT64 streamStopTime;
+    UINT64 streamStartTime;
+    STREAM_HANDLE streamHandle;
+    CHAR sampleDir[MAX_PATH_LEN + 1];
+} SampleCustomData, *PSampleCustomData;
+
+STATUS readFrameData(PFrame pFrame, PCHAR frameFilePath)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    UINT64 size;
+
+    CHK(pFrame != NULL && frameFilePath != NULL, STATUS_NULL_ARG);
+
+    // Get the size and read into frame
+    CHK_STATUS(readFile(frameFilePath, TRUE, NULL, &size));
+    CHK_STATUS(readFile(frameFilePath, TRUE, pFrame->frameData, &size));
+
+    pFrame->size = (UINT32) size;
+
+CleanUp:
+
+    return retStatus;
+}
+
+PVOID putVideoFrameRoutine(PVOID args)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    PSampleCustomData data = (PSampleCustomData) args;
+    Frame frame;
+    UINT32 fileIndex = 0, frameSize;
+    CHAR filePath[MAX_PATH_LEN + 1];
+    STATUS status;
+    BYTE frameBuffer[200000]; // Assuming this is enough
+    UINT64 runningTime;
+
+    CHK(data != NULL, STATUS_NULL_ARG);
+
+    frame.frameData = frameBuffer;
+    frame.version = FRAME_CURRENT_VERSION;
+    frame.trackId = DEFAULT_VIDEO_TRACK_ID;
+    frame.duration = 0;
+    frame.decodingTs = 0; // relative time mode
+    frame.presentationTs = 0;
+    frame.index = 0;
+
+    // video track is used to mark new fragment. A new fragment is generated for every frame with FRAME_FLAG_KEY_FRAME
+    frame.flags = fileIndex % DEFAULT_KEY_FRAME_INTERVAL == 0 ? FRAME_FLAG_KEY_FRAME : FRAME_FLAG_NONE;
+    SNPRINTF(filePath, MAX_PATH_LEN, "%s/h264SampleFrames/frame-%03d.h264", data->sampleDir, fileIndex + 1);
+    CHK_STATUS(readFrameData(&frame, filePath));
+
+    while (defaultGetTime() < data->streamStopTime) {
+        status = putKinesisVideoFrame(data->streamHandle, &frame);
+        ATOMIC_STORE_BOOL(&data->firstVideoFramePut, TRUE);
+        if (STATUS_FAILED(status)) {
+            DLOGD("putKinesisVideoFrame failed with 0x%08x", status);
+            status = STATUS_SUCCESS;
+        }
+
+        frame.presentationTs += SAMPLE_VIDEO_FRAME_DURATION;
+        frame.decodingTs = frame.presentationTs;
+        frame.index++;
+
+        fileIndex = (fileIndex + 1) % NUMBER_OF_H264_FRAME_FILES;
+        frame.flags = fileIndex % DEFAULT_KEY_FRAME_INTERVAL == 0 ? FRAME_FLAG_KEY_FRAME : FRAME_FLAG_NONE;
+        SNPRINTF(filePath, MAX_PATH_LEN, "%s/h264SampleFrames/frame-%03d.h264", data->sampleDir, fileIndex + 1);
+        CHK_STATUS(readFrameData(&frame, filePath));
+
+        // synchronize putKinesisVideoFrame to running time
+        runningTime = defaultGetTime() - data->streamStartTime;
```

Comment:
It's hard to understand what is going on and why our "simulated" frame media pipeline needs to synchronized. Suggest we use TimerQueue to fire the frames

Reply:
if we don't synchronize and just rely on thread sleep to simulate "realtime", audio and video timestamp gradually drifts apart and eventually the drift is too large and cause errors.